odo/backends/url.py

```diff
@@ -0,0 +1,129 @@
+from __future__ import print_function, division, absolute_import
+
+import os
+import uuid
+import zlib
+import re
+
+from contextlib import contextmanager
+
+import requests
+from toolz import memoize
+
+from .. import (discover, CSV, resource, append, convert, drop, Temp, JSON,
+                JSONLines, chunks)
+
+from multipledispatch import MDNotImplementedError
+
+from .text import TextFile
+
+from ..compatibility import urlparse
+from ..utils import tmpfile, ext, sample, filter_kwargs
+
+
+class _URL(object):
+    """ Parent class for data accessed through ``URLs``
+
+    Parameters
+    ----------
+
+    url : str
+        full url to file
+    chunk_size : int (default 1024)
+        Size of chunks streamed into memory
+    decode_unicode : bool (default False)
+        If True, content will be decoded using the best available encoding based on the response.
+
+     Examples
+    --------
+
+    >>> from odo import URL, CSV
+    >>> u = URL(CSV)('http://foo.com/myfile.csv')
+
+    Normally create through resource uris
+
+    >>> data = resource('http://foo.com/myfile.csv')
+    >>> data.url
+    'http://foo.com/myfile.csv'
+    >>> data.filename
+    'myfile.csv'
+    """
+    def __init__(self, url, chunk_size=1024, decode_unicode=False, *args,
+                 **kwargs):
+
+        self.subtype.__init__(self, url, *args, **kwargs)
+
+        self.url = url
+        self.chunk_size = chunk_size
+        self.decode_unicode = decode_unicode
+        self.filename = os.path.basename(urlparse(url).path)
+
+
+def URL(cls):
+    return type('URL(%s)' % cls.__name__, (_URL, cls), {'subtype': cls})
+
+URL.__doc__ = _URL.__doc__
+URL = memoize(URL) # what's happening here
```

Comment:
Instead of returning a new type, memoize will cache the result of the call to URL based on the hash of the inputs to each call. If the args are in the dictionary, return the values keyed by those args otherwise call the actual function.